elements/dash-tab-panel/dash-tab-panel.js

```diff
@@ -4,174 +4,217 @@
+     */
+    disconnectedCallback() {
+      this.removeEventListener('keydown', this._onKeyDown);
+      this.removeEventListener('click', this._onClick);
     }
-    // Select the new tab, that has been determined in the switch-case.
-    this._selectTab(newTab);
-    // The browser might have some native functionality bound to the arrow
-    // keys. Prevent the browser from taking any actions.
-    event.preventDefault();
-  }
 
-  /**
-   * `_handleClick` handles clicks inside the tab panel.
-   */
-  _handleClick(event) {
-    // If the click was not targeted on a tab element itself,
-    // it was a click inside the a panel or on empty space. Nothing to do.
-    if (!event.target.getAttribute('aria-role', 'tab')) return;
-    // If it was on a tab element, though, select that tab.
-    this._selectTab(event.target);
+    /**
+     * `selectTab` marks the given tab as selected.
+     * Additionally, it unhides the panel corresponding to the given tab.
+     */
+    _selectTab(newTab) {
+      // Deselect all tabs and hide all panels.
+      this.reset();
+
+      const panels = Array.from(this.querySelectorAll('[role=tabpanel]'));
+      // Get the panel that corresponds to the given tab.
+      const newPanelId = newTab.getAttribute('aria-controls');
+      const newPanel = panels.find(panel => panel.id === newPanelId);
+      // If that panel doesn’t exist, abort.
+      if (!newPanel) throw new Error(`No panel with id ${newPanelId}`);
+
+      // Unhide the panel and mark the tab as active.
+      newPanel.setAttribute('aria-hidden', 'false');
+
+      newTab.setAttribute('aria-selected', 'true');
+      newTab.tabIndex = 0;
+      newTab.focus();
+    }
+
+    /**
+     * `_onKeyDown` handles key presses inside the tab panel.
+     */
+    _onKeyDown(event) {
+      // The browser might have some native functionality bound to the arrow
+      // keys. Prevent the browser from taking any actions.
+      event.preventDefault();
+      // If the keypress did not originate from a tab element itself,
+      // it was a keypress inside the a panel or on empty space. Nothing to do.
+      if (event.target.getAttribute('role') !== 'tab') return;
+      // Don’t handle modifier shortcuts typically used by assistive devices.
+      if (event.altKey) return;
+
+      let newIdx;
+      let newTab;
+      const tabs = Array.from(this.querySelectorAll('[role=tab]'));
+
+      // The switch-case will determine which tab should be marked as active
+      // depending on the key that was pressed.
+      switch (event.keyCode) {
+        // When up or left is pressed, the previous tab is selected. If the
+        // first tab is reached, the element loops around to the last tab.
+        case KEYCODE.LEFT:
+        case KEYCODE.UP:
+          // The element uses `findIndex` to find the index of the currently
+          // selected element and subtracts one to get the index of the previous
+          // element.
+          newIdx =
```

Comment:
It might be nice to turn these into helper methods like `_prevTab` and `_nextTab`. Then you could say _selectNextTab or _selectPrevTab inside of these switch statements. That's what I did in the radio-group. That way we could start to zero in on some consistent patterns since I imagine the big keyboard switch/case will be pretty common in most of our elements